File: python-version/LC2131/longest_palindrome.py
class Solution:
    @staticmethod
    def longestPalindrome(words: list[str]) -> int:
        mp = {}
        for word in words:
            mp[word] = mp.get(word, 0) + 1
        st = set()
        odd = False
        ans = 0

        for word, cnt in mp.items():
            reversed_word = word[::-1]
            if word == reversed_word:
                if cnt & 1 == 1:
                    odd = True
                ans += (cnt // 2) * 2 * 2
            elif word not in st:
                st.add(word)
                st.add(reversed_word)
                ans += min(mp.get(word, 0), mp.get(reversed_word, 0)) * 2 * 2
        if odd:
            ans += 2
        return ans
        

File: go-version/leetcode/LC2131/longest_palindrome.go
package LC2131

/**
 * @Author: cutejiuge cutejiuge@163.com
 * @Date: 2025/5/25 下午6:52
 * @FilePath: longest_palindrome
 * @Description:
 */

func longestPalindrome(words []string) int {
	mp := make(map[string]int)
	for _, word := range words {
		mp[word]++
	}
	st := make(map[string]struct{})
	ans := 0
	odd := false

	for word, cnt := range mp {
		reversedWord := string(word[1]) + string(word[0])
		if reversedWord == word {
			if cnt&1 == 1 {
				odd = true
			}
			ans += (cnt / 2) * 2 * 2
		} else if _, ok := st[reversedWord]; !ok {
			st[word] = struct{}{}
			st[reversedWord] = struct{}{}
			ans += min(mp[word], mp[reversedWord]) * 2 * 2
		}
	}
	if odd {
		ans += 2
	}
	return ans
}
